CPU improvements and fixes

Holladiewal opened the pull request:
This PR includes several edits:
 - ~~There was a typo in the big match{} in run_opcode~~ This was already fixed in the `wip` branch
 - #5 increased the PC offset in JSR by 1 (to pc + 3). While this fixed the symptoms of RTS returning to an operand rather than an opcode, the error was in RTS not incrementing the pc by 1 after pulling it from the stack. Fixed now.
 - Stack access was all over the code and warranted some helpers. Added them. Also removed CPU::pull_pc() as it degenerated to `.pc = pull_word()` and it was used only in RTS and RTI, with RTS needing to increment the PC anyways.
 - BRK was pushing the current PC on the stack, not PC + 2. This would have resulted in a BRK loop, since RTI returns to the pushed PC. 
 - The addressing modes (and/or read_memory helpers) lack wrapping support, which should be easy to implement:
    - [x] zero page, zero page x, zero page y
    - [x] indirect zero page x
    - [x] indirect zero page y
    - [x] absolute 
    - [x] absolute x, absolute y
- JMP is a fun one when the address for indirection starts on the last byte of the page, as it does not continue the read on the next page, but wraps around on the current page
- Added missing addressing mode helpers, and used them in places not previously used.

Holladiewal committed:
CPU: Fix program_counter offset in RTS and JSR

PR #5 changed JSR to push pc + 3 to the stack.
This fixed the symptoms of JSR not returning to the right address,
but RTS is expecting the pushed PC address to be one before the opcode
to return too, as it will itself increment the PC after it is done.

Some programs may abuse this RTS behaviour (see http://www.6502.org/tutorials/6502opcodes.html#RTS)

Holladiewal committed:
CPU: Use helpers for stack access

REMOVES CPU::pull_pc()

We access the stack in many different places directly.
We can make the code cleaner and easier by using helpers.
CPU now has pull_word(), pull_byte(), push_byte() and push_word()
for (direct) stack interaction.

push_status() and pull_status() remain. pull_pc() is no longer
warranted as it only consisted of `.pc = pull_word()` which for
cleaner code can be a part of rts() and rti(), especially since
rts() needs to touch the PC anyways.

Holladiewal committed:
CPU: Fix BRK opcode

The BRK handler erroneously pushed the current PC onto the stack, not PC + 2.
This might have happened since we slightly deviate from the 6502 in regards to PC
position updates.
This also reorders some of the code to be closer in emulation to the 6502.

Holladiewal committed:
CPU: Add wrapping behaviour to indirect_zero_page_y()

If an address goes beyond 0xFFFF, it just wraps around. In our emulator,
it instead trips a run-time check within jakt, as our result would go past
the range of u16. For now, we do the address calculation in a u32, mask
everything out, that is not in the lower half, and cast back to u16.

Wrapping needs to be added to the other address reads too, which will
be done in a later commit

Holladiewal committed:
CPU: Little fixes

I messed some bits up during on of the earlier commits, lets just
fix that, so everything works as expected

pedro-ximenez commented on src/cpu.jakt:
```diff
@@ -1474,13 +1474,38 @@ class CPU {
+    
+    function pull_byte(mut this) -> u8 {
+        .s += 1
+        return .system.read_byte(address: (0x100u16 + .s as! u16))
+    }
+    
+    function pull_word(mut this) -> u16 {
+        .s += 1
+        let value = .system.read_word(address: 0x100u16 + .s as! u16)
+        .s += 1
+        return value
```
Would it be better to implement this in an analogous way to `.push_word`?

pedro-ximenez commented on src/cpu.jakt:
```diff
@@ -1474,13 +1474,38 @@ class CPU {
+    }
+
+    function push_word(mut this, value: u16) {       
+
```
Nit: blank line

pedro-ximenez commented on src/cpu.jakt:
```diff
@@ -1474,13 +1474,38 @@ class CPU {
+        .push_byte(value: first_byte)
+
+        let second_byte = (value & 0x00ff) as! u8
+        .push_byte(value: second_byte)
```
This core is incorrect. Using logical `&&` and not bit-wise`&` and there's a missing right shift.

To support pushing PC for `JSR`, it should be:
```
let hi_byte = ((value & 0xFF00) >> 8) as! u8;
let lo_byte = (value & 0xFF) as! u8;

.push_byte(value: hi_byte);
.push_byte(value: lo_byte);
```

To avoid a run-time check, it might be preferable to use `as_truncated<u8>()` instead of `as! u8`. @jntrnr, do you have a preference?

Holladiewal replied:
> This core is incorrect. Using logical && and not bit-wise& and there's a missing right shift.

This was corrected by the fixes commit included as part of the PR. Masking the upper byte is also not needed when rightshifting, as a u16 will have nothing else after the shift anyways. This should also prevent us exceeding the u8 range.

pedro-ximenez reviewed (changes requested):
This can be better written as:
```
let new_address = unchecked_add<u16>(pre_index, .y)
```

Nit: Extra blank lines.

jntrnr commented:
@Holladiewal - thanks for the PR, much appreciated

I've got a lot of changes getting ready to land after the videos go live this week. If you want to have a look, they're in the `wip` branch on the repo.

If you don't want spoilers, no worries. If you don't mind, can you rebase your changes and land them against the `wip` branch? If you'd rather wait, I can message back here once the branch has landed.

jntrnr commented:
@Holladiewal I think you might have included my `wip` branch in your PR against `main`. Instead, your changes to `wip` should go against the `wip` branch.

Holladiewal committed:
CPU: Add more wrapping support to adressing mode helpers

All addressing modes that are indexed, can exceed the range of u8/u16.
The 6502 deals with that by wrapping around, basically ignoring the overflown
bits. Jakt wants us to be safe and checks at runtime, that our calculations don't
exceed the range of the type. Since we know what we are doing, and C++ should do the
truncating for us when converting the type down after promotion, we use unchecked_add
to take advantage of that and dont get a runtime abort.

Holladiewal committed:
CPU: Change pull_word helper to use pull_byte

jntrnr commented:
Great, thanks. Are you wanting to continue adding more fixes to this or do you want to land it as it?

Holladiewal committed:
CPU: Add wrapping support to the rest of addressing modes

unchecked_add is our friend here, allowing wrapping to be "easily" implemented

Holladiewal committed:
CPU: Make JMP be more accurate.

JMP is straight forward unless you have to cross page boundaries during
the indirect read, because the 6502 will latch the high byte and wrap
around on the current page instead of continuing on the next page.
While you should never read the indirect vector from the last byte of
the page to save your sanity, nothing stops you from doing that and
nestest has a test for this too!

Holladiewal commented:
> Great, thanks. Are you wanting to continue adding more fixes to this or do you want to land it as it?

There is one thing I want to look into that would fit this PR without making it too big. When I'm happy I'll take the PR out of draft.

Holladiewal committed:
CPU: Finish addressing mode helpers

We were missing some helpers for some addressing modes, added them.
Also all functions should now use the helpers.

Holladiewal committed:
CPU: code cleanup

Holladiewal committed:
CPU: Fix mistakes

jntrnr commented:
In testing this after I landed both the background rendering and these CPU changes, I noticed that this breaks nestest. If I revert this PR on my local branch and just use the CPU before it, nestest is able to render the background properly.

I'm trying to debug now to see why fixes break nestest, but figured I'd mention it here in case you could spot the issue

sophiajt commented on src/cpu.jakt:
```diff
@@ -1258,63 +1285,49 @@ class CPU {
-                let address = .system.read_byte(address: arg_address)
-                .system.write_byte(address: address as! u16, value: .a)
+                
+                .zero_page_set(value: .a)
```
ahh, here's the problem (or one of them). This doesn't work since we've already modified the pc. I do the same bug in one of the videos and have to fix it.

Holladiewal replied:
Oh. Yeah, you are correct, that won't work. It's probably bugged in all the places I changed it then, since I just replaced the read with the helper, and PC increments are (usually) all at the start of opcodes.

sophiajt commented on src/cpu.jakt:
```diff
@@ -1696,17 +1712,13 @@ class CPU {
-                let address = .system.read_byte(address: arg_address)
-                let value = .system.read_byte(address: address as! u16)
-                yield value
+                yield .zero_page()
```
same bug here
